### PR TITLE
raft-leases: Handle failures writing lease state to the database

### PR DESCRIPTION
**Pull request by babbageclunk**
## Description of change

If the raftlease trapdoor notices that the DB leaseholder state is out of sync, it updates the database with the right holder. This is safe to do because trapdoor is only ever called when we've checked the current state in raft is good.

To avoid load on the DB for something that should only happen occasionally, we only check the leaseholder state on the second attempt at building+running a leadership transaction. This requires that `lease.Token.Check`, `lease.Trapdoor` and `leadership.Token.Check` get passed in the attempt (we can't keep it as closure state because the trapdoor is recreated each time).

The lease manager manifold now needs the state worker, because the trapdoor factory needs a *state.State so that it can update leaseholders when needed.

## QA steps

* Compile this with the wrench in notifyTarget.Claimed uncommented. 
* Bootstrap a controller and enable HA.
* Run `echo fail-claim > /var/lib/juju/wrench/raftlease-notifytarget` on each controller to prevent the DB from being updated when a lease is claimed in raft.
* Watch the controller debug-log for errors.
* Deploy leadership-flex (https://github.com/mitechie/leadership-charm) - it sets leadership settings when a unit is elected leader.
* When the first unit is deployed you should see the wrench error in the controller log (on the raft leader) indicating that we couldn't update the database with the new lease.
* The new unit should still start up successfully, including setting leadership settings - its state will be "ready to go". No hooks should fail, and status should indicate it's the leader.
* The database should contain the leaseholder record for leadership-flex.
* Add another unit of leadership-flex.
* Once it's running, remove the original unit.
* When the lease expires the new unit should claim the leadership, no hooks fail.
* The controller error log will show the claim couldn't be written to the db when the raft change occurred (because of the wrench).
* Status will indicate the new unit is the leader, and the database will indicate the new lease holder.

## Documentation changes

None

## Bug reference

Fixes https://bugs.launchpad.net/juju/+bug/1810331

### CODE REVIEW

**Comment by jameinel** [comment]
I should clarify that if we are going to have the code to talk to mongo in the raft code, then we should just do that. It feels like the wrong abstraction in the long term, but in the short term we can still make things better with what we have. And for that, I think this patch is good.

File: state/raftlease/target.go [hunk 1]
```diff
@@ -203,25 +212,48 @@ func (t *notifyTarget) Expired(key lease.Key) {
...
+				ops = assertionOps
+			} else {
+				// If the assertion failed it may be because a claim
+				// notify failed in the past due to the DB not being
+				// available. Check whether we need to sync the lease
+				// holder - this is safe to do because raft is the
+				// arbiter of who really holds the lease, and we check
+				// that the lease is held in buildTxnWithLeadership
+				// each time before collecting the assertion ops.
+				docId := leaseHolderDocId(key.Namespace, key.ModelUUID, key.Lease)
```
[the remark this answers is not in the document]
Comment (babbageclunk) [reply state/raftlease/target.go]: Yeah, I think you're right about this - I'll update it to make the change immediately.

File: state/raftlease/target.go [hunk 2]
```diff
@@ -203,25 +212,48 @@ func (t *notifyTarget) Expired(key lease.Key) {
...
-				key.Namespace,
-				key.ModelUUID,
-				key.Lease,
-			),
-			Assert: bson.M{
-				fieldHolder: holder,
-			},
-		}
-		return func(out interface{}) error {
+		return func(attempt int, out interface{}) error {
```
[the remark this answers is not in the document]
Comment (babbageclunk) [reply state/raftlease/target.go]: I think this is a misunderstanding - let's discuss when you're around later on today?

**Comment by babbageclunk** [comment]
$$merge$$

**Pull request merged**